### PR TITLE
Quasar chain uprgade

### PR DESCRIPTION
**Pull request by arhamchordia**
## 1. Overview

Chain upgrade from sdk45 to sdk47

## 2. Implementation details

Changes made in this PR are mostly on adding few new modules into play and upgrading versions of old dependencies : 
1. Wasm bumped from v31 to v45
2. IBC from v4 to v7
3. Wasmvm from 1.2 to 1.5
4. IBC was light client added
5. Cosmos sdk from v45 to v47.12
6. Async ICQ from v4 to v7
7. Tendermint to cometbft
8. and so on... 

## 3. How to test/use

Testing can be done by running : 
1. Compile v1.0.0 or download from tagged versions as `quasarnodedv1`
2. Change directory to `demos/upgrade-handler/v2.0.0`
3. Run upgrade_test.sh. 

Upgrade test will run the old binary with relayer setup and some IBC connections to Osmosis chain. 
And then it will perform some pre upgrade transactions and then post the upgrade proposal. 
Once the desired block height is reached then the setup will automatically stop the old chain and start with new binary. 
New binary once started, it will be tested using the post_upgrade actions on chain. 

## 4. Checklist

- [ ] Does the Readme need to be updated?

### CODE REVIEW

File: app/keepers/keepers.go [hunk 1]
```diff
@@ -76,5 +132,455 @@ type AppKeepers struct {
...
+
+	// AUDIT CHECK IS THIS TYPE ASSERTION FOR TYPE CASTING INTERFACE TO STRUCT SAFE?
+	tmpBankBaseKeeper := appKeepers.BankKeeper.(bankkeeper.BaseKeeper)
+
+	wasmOpts = append(owasm.RegisterCustomPlugins(appKeepers.QOracleKeeper, &tmpBankBaseKeeper, callback), wasmOpts...)
+	wasmOpts = append(tfbindings.RegisterCustomPlugins(&tmpBankBaseKeeper, &appKeepers.TfKeeper), wasmOpts...)
+
+	// The last arguments can contain custom message handlers, and custom query handlers,
+	// if we want to allow any custom callbacks
+	supportedFeatures := "iterator,staking,stargate"
```
Comment (akure) [inline app/keepers/keepers.go:477]: @arhamchordia 
cosmwasm_1_1,cosmwasm_1_2,cosmwasm_1_4"  are missing in the supported feature.

File: cmd/quasarnoded/main.go [hunk 2]
```diff
@@ -10,11 +10,13 @@ import (
 	"github.com/quasarlabs/quasarnode/cmd/quasarnoded/cmd"
 )
 
+var ENV_PREXIX = "QUASARNODED"
```
Comment (akure) [inline cmd/quasarnoded/main.go:13]: Could yoy please re-check this ENV variable usage that it doe not break anything.?

File: app/keepers/keepers.go (hunk 1 above)
Comment (arhamchordia) [reply app/keepers/keepers.go:477]: Added

File: cmd/quasarnoded/main.go (hunk 2 above)
Comment (arhamchordia) [reply cmd/quasarnoded/main.go:13]: This env prefix is used for : 
// SetEnvPrefix defines a prefix that ENVIRONMENT variables will use.
// E.g. if your prefix is "spf", the env registry will look for env
// variables that start with "SPF_".

Comment (arhamchordia) [reply cmd/quasarnoded/main.go:13]: We can leave it empty if needed. But it does not do much.

**Review by ajansari95 — CHANGES_REQUESTED** [review]
A thorough cleanup of the PR is required. Points that need to be addressed include:

- A lot of commented code.
- Inconsistent import grouping.
- TODOs , if addressed, should be removed. Additionally, we have at numerous placesmentioned   we need to recheck if the fix works was it rechecked?if yes the comment should be removed.

I believe our CI is very weak; this PR shouldn't have passed any lint checks. We need to specify strict linting rules.

File: x/epochs/client/cli/cli_test.go [hunk 3]
```diff
@@ -1,115 +1,104 @@
...
+//	cmd := cli.GetCmdEpochsInfos()
+//
+//	out, err := clitestutil.ExecTestCLICmd(clientCtx, cmd, common)
+//	require.NoError(t, err, out.String())
+//
+//	var resp types.QueryEpochsInfoResponse
+//	err = clientCtx.Codec.UnmarshalJSON(out.Bytes(), &resp)
+//	require.NoError(t, err)
+//	require.Equal(t, 4, len(resp.Epochs))
+//}
```
[the remark this answers is not in the document]
Comment (arhamchordia) [reply x/epochs/client/cli/cli_test.go:104]: Nope. Some issue in sdk which might be addressed later.

File: proto/osmosis/gamm/v1beta1/query.proto [hunk 4]
```diff
@@ -73,7 +73,8 @@ message QueryPoolRequest {
   uint64 pool_id = 1 [ (gogoproto.moretags) = "yaml:\"pool_id\"" ];
 }
 message QueryPoolResponse {
-  google.protobuf.Any pool = 1 [ (cosmos_proto.accepts_interface) = "PoolI" ];
+  // google.protobuf.Any pool = 1 [ (cosmos_proto.accepts_interface) = "PoolI" ];
```
Comment (magiodev) [inline proto/osmosis/gamm/v1beta1/query.proto:76]: comments here and below

File: app/app.go [hunk 5]
```diff
@@ -1,145 +1,59 @@
...
-	qvestingmodulekeeper "github.com/quasarlabs/quasarnode/x/qvesting/keeper"
-	qvestingmoduletypes "github.com/quasarlabs/quasarnode/x/qvesting/types"
-	tfmodule "github.com/quasarlabs/quasarnode/x/tokenfactory"
-	tfbindings "github.com/quasarlabs/quasarnode/x/tokenfactory/bindings"
-	tfkeeper "github.com/quasarlabs/quasarnode/x/tokenfactory/keeper"
-	tftypes "github.com/quasarlabs/quasarnode/x/tokenfactory/types"
-	// this line is used by starport scaffolding # stargate/app/moduleImport
 )
 
 const (
```
[the remark this answers is not in the document]
Comment (arhamchordia) [reply app/app.go:59]: Keeping it for sdk v50.

File: app/app.go [hunk 6]
```diff
@@ -153,55 +67,29 @@ var (
 	// are disabled.
 	WasmProposalsEnabled = "true"
 
-	// EnableSpecificWasmProposals, if set, must be comma-separated list of values
+	// EnableSpecificWasmProposals , if set, must be comma-separated list of values
 	// that are all a subset of "EnableAllProposals", which takes precedence over
 	// WasmProposalsEnabled.
 	//
 	// See: https://github.com/CosmWasm/wasmd/blob/02a54d33ff2c064f3539ae12d75d027d9c665f05/x/wasm/internal/types/proposal.go#L28-L34
 	EnableSpecificWasmProposals = ""
```
[the remark this answers is not in the document]
Comment (arhamchordia) [reply app/app.go:75]: removed.

File: x/qvesting/handler.go [hunk 7]
```diff
@@ -1,6 +1,7 @@
 package qvesting
 
 import (
+	errorsmod "cosmossdk.io/errors"
```
[the remark this answers is not in the document]
Comment (arhamchordia) [reply x/qvesting/handler.go:4]: used in sdk like this.

File: x/qvesting/keeper/msg_server.go [hunk 8]
```diff
@@ -2,6 +2,7 @@ package keeper
 
 import (
 	"context"
+	errorsmod "cosmossdk.io/errors"
```
[the remark this answers is not in the document]
Comment (arhamchordia) [reply x/qvesting/keeper/msg_server.go:5]: Used in sdk like this.

File: x/qvesting/module.go [hunk 9]
```diff
@@ -10,14 +10,14 @@ import (
...
-	abci "github.com/tendermint/tendermint/abci/types"
+	abci "github.com/cometbft/cometbft/abci/types"
 
 	"github.com/cosmos/cosmos-sdk/client"
 	"github.com/cosmos/cosmos-sdk/codec"
 	cdctypes "github.com/cosmos/cosmos-sdk/codec/types"
 	sdk "github.com/cosmos/cosmos-sdk/types"
 	"github.com/cosmos/cosmos-sdk/types/module"
-	"github.com/cosmos/cosmos-sdk/x/bank/client/rest"
+	// "github.com/cosmos/cosmos-sdk/x/bank/client/rest"
```
[the remark this answers is not in the document]
Comment (arhamchordia) [reply x/qvesting/module.go]: Can merge for now.

File: x/tokenfactory/bindings/query_plugin.go [hunk 10]
```diff
@@ -1,6 +1,7 @@
 package bindings
 
 import (
+	errorsmod "cosmossdk.io/errors"
```
[the remark this answers is not in the document]
Comment (arhamchordia) [reply x/tokenfactory/bindings/query_plugin.go:4]: SDK nomenclature.

File: x/qvesting/types/msgs.go [hunk 11]
```diff
@@ -1,6 +1,7 @@
 package types
 
 import (
+	errorsmod "cosmossdk.io/errors"
```
[the remark this answers is not in the document]
Comment (arhamchordia) [reply x/qvesting/types/msgs.go:4]: Yes. Wraf and Wrap are deprecated in sdk.

File: app/app.go [hunk 12]
```diff
@@ -962,12 +404,12 @@ func (app *App) InitChainer(ctx sdk.Context, req abci.RequestInitChain) abci.Res
 }
 
 // LoadHeight loads a particular height
-func (app *App) LoadHeight(height int64) error {
+func (app *QuasarApp) LoadHeight(height int64) error {
```
[the remark this answers is not in the document]
Comment (arhamchordia) [reply app/app.go:407]: Keeping it as QuasarApp as App seems generic. (struct from sdk is App)

File: app/keepers/keepers.go [hunk 13]
```diff
@@ -2,63 +2,117 @@ package keepers
...
+	ConsensusParamsKeeper *consensusparamkeeper.Keeper
 
 	// make scoped keepers public for test purposes
 	ScopedIBCKeeper           capabilitykeeper.ScopedKeeper
 	ScopedTransferKeeper      capabilitykeeper.ScopedKeeper
 	ScopedICAControllerKeeper capabilitykeeper.ScopedKeeper
 	ScopedICAHostKeeper       capabilitykeeper.ScopedKeeper
 	ScopedIntergammKeeper     capabilitykeeper.ScopedKeeper
-	ScopedQOracleKeeper       capabilitykeeper.ScopedKeeper
+	ScopedQOsmosisKeeper      capabilitykeeper.ScopedKeeper
```
[the remark this answers is not in the document]
Comment (arhamchordia) [reply app/keepers/keepers.go:99]: This is for getting prices from Osmosis about any token mentioned. Keeping it here till we see if there is any use case later on. Or else we can deprecated this later.

File: app/export.go [hunk 14]
```diff
@@ -15,7 +15,7 @@ import (
 
 // ExportAppStateAndValidators exports the state of the application for a genesis
 // file.
-func (app *App) ExportAppStateAndValidators(
+func (app *QuasarApp) ExportAppStateAndValidators(
```
[the remark this answers is not in the document]
Comment (arhamchordia) [reply app/export.go:17]: App is generic I feel

File: .github/workflows/build.yml [hunk 15]
```diff
@@ -40,7 +40,7 @@ jobs:
       - name: Setup Golang
         uses: actions/setup-go@v4
         with:
-          go-version: 1.20.7
+          go-version: 1.21
```
[the remark this answers is not in the document]
Comment (arhamchordia) [reply .github/workflows/build.yml:43]: Go.mod contains 1.21 rn.

File: Makefile [hunk 16]
```diff
@@ -176,11 +176,20 @@ go.sum: go.mod
...
 
 proto-all: proto-format proto-gen
-
+BUF_VERSION=1.26.1
+BUILDER_VERSION=0.13.5
 proto-gen:
 	@echo "Generating Protobuf files"
 	@sh ./scripts/protocgen.sh
 
+proto-gen-1:
```
[the remark this answers is not in the document]
Comment (arhamchordia) [reply Makefile:185]: Makefile undergoes overhaul in v50 upgrade.

**Review by magiodev — DISMISSED** [review]

File: app/export.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (magiodev) [reply app/export.go:17]: I still tend to say that generic is better. We do not need to brand our App inside the Quasar repository. Also inside the App context, we do not techincally need to know that App is QuasarApp, App is the SDK framework App we work with.

File: .github/workflows/build.yml (hunk 15 above)
[the remark this answers is not in the document]
Comment (magiodev) [reply .github/workflows/build.yml:43]: Yes I was meaning from a general point of view, here and related places. There is any motivation behind not bumping the whole go usage to the latest version?

**Review by ajansari95 — DISMISSED** [review]

**Review by akure — DISMISSED** [review]
Looks good.

**Review by ajansari95 — APPROVED** [review]

**Review by magiodev — APPROVED** [review]

**Review by akure — APPROVED** [review]

**Pull request merged**